File: leetcode/1001-2000/1401-1500/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/ReorderRoutesToMakeAllPathsLeadToTheCityZero.ts
function minReorder(n: number, connections: number[][]): number {
  const visited = new Array(n).fill(false);
  const graph = new Array(n).fill(null).map(
    (_) => ({ froms: [] as number[], tos: [] as number[] })
  )

  for (const [from, to] of connections) {
    graph[to].froms.push(from);
    graph[from].tos.push(to);
  }

  let totalReorders = 0;

  const visit = (i: number) => {
    if (visited[i]) {
      return;
    }

    visited[i] = true;

    const { froms, tos } = graph[i];

    for (const from of froms) {
      if (!visited[from]) {
        visit(from);
      }
    }

    for (const to of tos) {
      if (!visited[to]) {
        totalReorders++;
        visit(to);
      }
    }
  };

  visit(0);


  return totalReorders;
}
// function minReorder(n: number, connections: number[][]): number {
//   const visited = new Set([0]);
//
//   let totalReorders = 0;
//
//   while (connections.length > 0) {
//     const [from, to] = connections.shift()!;
//
//     if (visited.has(to)) {
//       visited.add(from);
//       continue;
//     }
//
//     if (visited.has(from)) {
//       visited.add(to);
//       totalReorders++;
//       continue;
//     }
//
//     connections.push([from, to]);
//   }
//
//   return totalReorders;
// };

console.log(minReorder(6, [[0, 1], [1, 3], [2, 3], [4, 0], [4, 5]]));
console.log(minReorder(5, [[1, 0], [1, 2], [3, 2], [3, 4]]));

File: leetcode/1001-2000/1401-1500/1461-1470/1466-reorder-routes-to-make-all-paths-lead-to-the-city-zero/ReorderRoutesToMakeAllPathsLeadToTheCityZero.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ReorderRoutesToMakeAllPathsLeadToTheCityZero {
  public static void main(String[] args) {
    int n = 6;
    int[][] connections = {
        { 0, 1 }, { 1, 3 }, { 2, 3 }, { 4, 0 }, { 4, 5 }
    };

    // console.log(minReorder(6, [[0, 1], [1, 3], [2, 3], [4, 0], [4, 5]]));
    System.out.println(new Solution().minReorder(n, connections));
  }
}

class Solution {
  int visit(int city, Map<Integer, List<Integer>> graph, boolean[] visited) {
    int totalReorders = 0;

    if (visited[city]) {
      return 0;
    }

    visited[city] = true;

    for (int k : graph.get(city)) {
      int neighbour = Math.abs(k);

      if (visited[neighbour]) {
        continue;
      }

      if (k < 0) {
        totalReorders++;
      }

      totalReorders += visit(neighbour, graph, visited);
    }

    return totalReorders;
  }

  public int minReorder(int n, int[][] connections) {
    Map<Integer, List<Integer>> graph = new HashMap<>();

    for (int[] connection : connections) {
      int from = connection[0];
      int to = connection[1];

      graph.computeIfAbsent(to, k -> new ArrayList<>()).add(from);
      graph.computeIfAbsent(from, k -> new ArrayList<>()).add(-to);
    }

    return visit(0, graph, new boolean[n]);
  }
}
